### DISCUSSION TITLE
Typing of RTK Query `createApi` return value in JetBrains WebStorm?

### DISCUSSION THREAD

**Discussion by joshuajung** [reduxjs/redux-toolkit#3397] [Q&A] [answered]
Hi everyone,

I'm evaluating a switch from VSCode to WebStorm for our frontend development. Everything works great so far, but I'm noticing that all our RTK Query typings are missing in code completion – see the detailed issue at https://youtrack.jetbrains.com/issue/WEB-60736/Return-type-of-createApi-from-redux-toolkit-is-not-correctly-inferred.

May I ask if any of you are using Webstorm for RTK Query development and can or cannot confirm this finding?

Thank you!

**Comment by phryneas** [comment, marked as answer]
Unfortunately, we cannot do much to support WebStorm until they start actually using TypeScript and not their own implementation. 

We have type tests that test the library against the last 5 versions of TypeScript to guarantee compatibility, but WebStorm uses some self-invented heuristic that often acts very similar to TypeScript, but oftentimes also completely goes off the rails and does stuff like you see there. 

I can really just recommend you to use an IDE for TypeScript development that actually uses the TypeScript language server (VSCode, vim, etc.).

**Reply by joshuajung** [reply]
Thanks a lot for the insights! I had no clue that WebStorm had its own TS implementation. I'll probably go back to VSCode for the time being.

**Reply by mitroc** [reply]
I just checked a piece of code against your snippets in YouTrack. Indeed, WS shows the API type as any, but it suggests correctly. Or am I missing something?
![ws](https://github.com/reduxjs/redux-toolkit/assets/25466575/eb039b3b-5e5e-4427-964d-519713c4a0fe)

**Reply by joshuajung** [reply]
> I just checked a piece of code against your snippets in YouTrack. Indeed, WS shows the API type as any, but it suggests correctly. Or am I missing something?

Oh, that's an interesting observation. Indeed, autocomplete seems to be working fine when being used inside the same file. When I import the file into another one with a lot of other unrelated TS code though, it only works in about 20% of the cases. Weird. I'll see if I can pin it down an further. Feel a bit like it's hitting some kind of timeout and then just giving up, as code completion is rather slow for this file in VSCode as well.
